compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -800,137 +818,187 @@ trait Implicits { self: Typer =>
-        val rt1 = alt1.ref.widen.resultType
-        val rt2 = alt2.ref.widen.resultType
+          val rpt = pt.resultType
+          val rt1 = alt1.ref.widen.resultType
+          val rt2 = alt2.ref.widen.resultType
         if (isNumeric(rpt) && isNumeric(rt1) && isNumeric(rt2))
-          if (isProperSubType(rt1, rt2)) alt1
-          else if (isProperSubType(rt2, rt1)) alt2
-          else NoImplicitMatches
-        else NoImplicitMatches
+          if (isProperSubType(rt1, rt2)) alt2
+          else if (isProperSubType(rt2, rt1)) alt1
+          else NoMatchingImplicitsFailure
+        else NoMatchingImplicitsFailure
       }
+      */
 
-      /** Convert a (possibly empty) list of search successes into a single search result */
-      def condense(hits: List[SearchSuccess]): SearchResult = hits match {
-        case best :: alts =>
-          alts.find(alt =>
-            ctx.typerState.test(isAsGood(alt.ref, best.ref, alt.level, best.level))) match {
-            case Some(alt) =>
-              typr.println(i"ambiguous implicits for $pt: ${best.ref} @ ${best.level}, ${alt.ref} @ ${alt.level}")
-            /* !!! DEBUG
-              println(i"ambiguous refs: ${hits map (_.ref) map (_.show) mkString ", "}")
-              isAsGood(best.ref, alt.ref, explain = true)(ctx.fresh.withExploreTyperState)
-            */
-              numericValueTieBreak(best, alt) match {
-                case eliminated: SearchSuccess => condense(hits.filter(_ ne eliminated))
-                case _ => new AmbiguousImplicits(best.ref, alt.ref, pt, argument)
+      /** If `alt1` is also a search success, try to disambiguate as follows:
+       *    - If alt2 is preferred over alt1, pick alt2, otherwise return an
+       *      ambiguous implicits error.
+       */
+      def disambiguate(alt1: SearchResult, alt2: SearchSuccess) = alt1 match {
+        case alt1: SearchSuccess =>
+          val diff = compareCandidate(alt1, alt2.ref, alt2.level)
+          assert(diff <= 0)
+          if (diff < 0) alt2
+          else
+            // numericValueTypeBreak(alt1, alt2) recoverWith
+            SearchFailure(new AmbiguousImplicits(alt1, alt2, pt, argument))
+        case _: SearchFailure => alt2
+      }
+
+      def healAmbiguous(pending: List[Candidate], fail: SearchFailure) = {
+        val ambi = fail.reason.asInstanceOf[AmbiguousImplicits]
+        val newPending = pending.filter(cand =>
+          compareCandidate(ambi.alt1, cand.ref, cand.level) < 0 &&
+          compareCandidate(ambi.alt2, cand.ref, cand.level) < 0)
+        rank(newPending, fail, Nil).recoverWith(_ => fail)
+      }
+
+      def rank(pending: List[Candidate], found: SearchResult, rfailures: List[SearchFailure]): SearchResult = pending match {
+        case cand :: pending1 =>
+          tryImplicit(cand) match {
+            case fail: SearchFailure =>
+              if (isNot)
+                SearchSuccess(ref(defn.Not_value), defn.Not_value.termRef, 0)(ctx.typerState)
```

Comment:
I don't have a minimized example to reproduce this but there is something wrong here: The `ctx` used here comes from the `ImplicitSearch` class implicit parameter. There is no guarantee that the TyperState of this Context is committable (this can be checked by adding `assert(ctx.typerState.isCommittable)`). This is bad because in `inferImplicit` we do `result.tstate.commit()` which crashes when the TyperState is not committable. This is only a problem here because the other call to `SearchSuccess` is done in `typedImplicit` which takes its own `implicit ctx` (the only call to `typedImplicit` is `typedImplicit(cand)(nestedContext().setNewTyperState().setSearchHistory(history))`, the `setNewTyperState()` ensures the state is always committable).

I found this out while trying to get the usage of ambiguous implicits in `shims` to work, see https://github.com/smarter/shims/commits/dotty-not

Reply:
Well spotted! Yes, we need to give it a nested committable typerstate.